server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxNameEscaper.java

```diff
@@ -0,0 +1,126 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.james.util.OptionalConverter;
+
+import com.google.common.base.Preconditions;
+import com.google.common.escape.CharEscaper;
+
+public class MailboxNameEscaper {
+
+    public static final CharEscaper CHAR_ESCAPER = new CharEscaper() {
+        @Override
+        protected char[] escape(char c) {
+            if (c == '\\') {
+                return "\\\\".toCharArray();
+            }
+            if (c == '/') {
+                return "//".toCharArray();
+            }
+            if (c == '.') {
+                return "\\/".toCharArray();
+            }
+            return null;
+        }
+    };
+
+    public static String escape(String string) {
+        Preconditions.checkNotNull(string);
+
+        return CHAR_ESCAPER.escape(string);
+    }
+
+    public static String unescape(String string) {
+        Preconditions.checkNotNull(string);
+
+        List<CharParsingResult> charParsingResults = new ArrayList<>(string.length());
+        CharParsingResult lastEscapeChar = empty();
+        for (char c : string.toCharArray()) {
+            lastEscapeChar = parse(lastEscapeChar, c);
+            charParsingResults.add(lastEscapeChar);
+        }
+        return charParsingResults.stream()
+            .flatMap(charParsingResult -> OptionalConverter.toStream(charParsingResult.result))
+            .map(Object::toString)
+            .collect(Collectors.joining());
+    }
+
+    static CharParsingResult empty() {
+        return new CharParsingResult(Optional.empty(), Optional.empty());
+    }
+
+    static CharParsingResult escapeChar(char c) {
+        Preconditions.checkArgument(isEscapeChar(c));
+        return new CharParsingResult(Optional.of(c), Optional.empty());
+    }
+
+    private static boolean isEscapeChar(char c) {
+        return c == '\\' || c == '/';
+    }
+
+    static CharParsingResult normalChar(char c) {
+        return new CharParsingResult(Optional.empty(), Optional.of(c));
+    }
+
+    static CharParsingResult escapedChar(char escapeChar, char escapedChar) {
+        if (escapeChar == '\\') {
```

Comment:
I don't see how to deduplicate more logic in here. 

From an analitical point of view : 
 - You need to check your context 
 - Then to check your current value

I will see if I can encapsulate this decision making as part of the context (and if it looks better)